File: LLM-functionalist-translations/X-grok/3-think/py-translations/chat-code.py
###given as a file named: stream_processing.py

import sys
import ipaddress
from typing import Dict, List, Tuple, Callable, Optional
from io import TextIOWrapper

# OpResult variants
class OpResult:
    pass

class Float(OpResult):
    def __init__(self, value: float):
        self.value = value

class Int(OpResult):
    def __init__(self, value: int):
        self.value = value

class IPv4(OpResult):
    def __init__(self, value: ipaddress.IPv4Address):
        self.value = value

class MAC(OpResult):
    def __init__(self, value: bytes):
        self.value = value

class Empty(OpResult):
    pass

# Tuple is a dictionary from strings to OpResult
TupleType = Dict[str, OpResult]

# Operator base class
class Operator:
    def next(self, tup: TupleType) -> None:
        raise NotImplementedError

    def reset(self, tup: TupleType) -> None:
        raise NotImplementedError

# Conversion utilities
def string_of_mac(buf: bytes) -> str:
    return ":".join(f"{byte:02x}" for byte in buf[:6])

def tcp_flags_to_strings(flags: int) -> str:
    flag_names = {
        "FIN": 1 << 0,
        "SYN": 1 << 1,
        "RST": 1 << 2,
        "PSH": 1 << 3,
        "ACK": 1 << 4,
        "URG": 1 << 5,
        "ECE": 1 << 6,
        "CWR": 1 << 7,
    }
    return "|".join(name for name, value in flag_names.items() if flags & value == value)

def int_of_op_result(input: OpResult) -> int:
    if isinstance(input, Int):
        return input.value
    raise ValueError("Trying to extract int from non-int result")

def float_of_op_result(input: OpResult) -> float:
    if isinstance(input, Float):
        return input.value
    raise ValueError("Trying to extract float from non-float result")

def string_of_op_result(input: OpResult) -> str:
    if isinstance(input, Float):
        return str(input.value)
    elif isinstance(input, Int):
        return str(input.value)
    elif isinstance(input, IPv4):
        return str(input.value)
    elif isinstance(input, MAC):
        return string_of_mac(input.value)
    elif isinstance(input, Empty):
        return "Empty"
    raise ValueError("Unknown OpResult type")

def string_of_tuple(input_tuple: TupleType) -> str:
    return ", ".join(f'"{key}" => {string_of_op_result(value)}' for key, value in input_tuple.items())

def tuple_of_list(tup_list: List[Tuple[str, OpResult]]) -> TupleType:
    return dict(tup_list)

def dump_tuple(outc: TextIOWrapper, tup: TupleType) -> None:
    print(string_of_tuple(tup), file=outc)

def lookup_int(key: str, tup: TupleType) -> int:
    return int_of_op_result(tup[key])

def lookup_float(key: str, tup: TupleType) -> float:
    return float_of_op_result(tup[key])

# Built-in operator definitions
INIT_TABLE_SIZE = 10000

class DumpTupleOperator(Operator):
    def __init__(self, outc: TextIOWrapper, show_reset: bool = False):
        self.outc = outc
        self.show_reset = show_reset

    def next(self, tup: TupleType):
        dump_tuple(self.outc, tup)

    def reset(self, tup: TupleType):
        if self.show_reset:
            dump_tuple(self.outc, tup)
            print("[reset]", file=self.outc)

class DumpAsCsvOperator(Operator):
    def __init__(self, outc: TextIOWrapper, static_field: Optional[Tuple[str, str]] = None, header: bool = True):
        self.outc = outc
        self.static_field = static_field
        self.first = header

    def next(self, tup: TupleType):
        if self.first:
            if self.static_field:
                print(f"{self.static_field[0]},", end="", file=self.outc)
            print(",".join(tup.keys()), file=self.outc)
            self.first = False
        if self.static_field:
            print(f"{self.static_field[1]},", end="", file=self.outc)
        print(",".join(string_of_op_result(value) for value in tup.values()), file=self.outc)

    def reset(self, tup: TupleType):
        pass

class DumpWaltsCsvOperator(Operator):
    def __init__(self, filename: str):
        self.filename = filename
        self.outc = sys.stdout
        self.first = True

    def next(self, tup: TupleType):
        if self.first:
            self.outc = open(self.filename, 'w')
            self.first = False
        print(f"{tup['src_ip']},{tup['dst_ip']},{tup['src_l4_port']},{tup['dst_l4_port']},"
              f"{tup['packet_count']},{tup['byte_count']},{tup['epoch_id']}", file=self.outc)

    def reset(self, tup: TupleType):
        pass

def get_ip_or_zero(input: str) -> OpResult:
    if input == "0":
        return Int(0)
    return IPv4(ipaddress.IPv4Address(input))

class EpochOperator(Operator):
    def __init__(self, epoch_width: float, key_out: str, next_op: Operator):
        self.epoch_width = epoch_width
        self.key_out = key_out
        self.next_op = next_op
        self.epoch_boundary = 0.0
        self.eid = 0

    def next(self, tup: TupleType):
        time = lookup_float("time", tup)
        if self.epoch_boundary == 0.0:
            self.epoch_boundary = time + self.epoch_width
        elif time >= self.epoch_boundary:
            while time >= self.epoch_boundary:
                self.next_op.reset({self.key_out: Int(self.eid)})
                self.epoch_boundary += self.epoch_width
                self.eid += 1
        self.next_op.next({**tup, self.key_out: Int(self.eid)})

    def reset(self, tup: TupleType):
        self.next_op.reset({self.key_out: Int(self.eid)})
        self.epoch_boundary = 0.0
        self.eid = 0

class FilterOperator(Operator):
    def __init__(self, f: Callable[[TupleType], bool], next_op: Operator):
        self.f = f
        self.next_op = next_op

    def next(self, tup: TupleType):
        if self.f(tup):
            self.next_op.next(tup)

    def reset(self, tup: TupleType):
        self.next_op.reset(tup)

def key_geq_int(key: str, threshold: int) -> Callable[[TupleType], bool]:
    def f(tup: TupleType) -> bool:
        return lookup_int(key, tup) >= threshold
    return f

def get_mapped_int(key: str, tup: TupleType) -> int:
    return lookup_int(key, tup)

def get_mapped_float(key: str, tup: TupleType) -> float:
    return lookup_float(key, tup)

class MapOperator(Operator):
    def __init__(self, f: Callable[[TupleType], TupleType], next_op: Operator):
        self.f = f
        self.next_op = next_op

    def next(self, tup: TupleType):
        self.next_op.next(self.f(tup))

    def reset(self, tup: TupleType):
        self.next_op.reset(tup)

def make_hashable(key: TupleType) -> tuple:
    return tuple(sorted((k, v.value) for k, v in key.items()))

class GroupByOperator(Operator):
    def __init__(self, groupby: Callable[[TupleType], TupleType], reduce: Callable[[OpResult, TupleType], OpResult],
                 out_key: str, next_op: Operator):
        self.groupby = groupby
        self.reduce = reduce
        self.out_key = out_key
        self.next_op = next_op
        self.h_tbl = {}  # Dict[tuple, Tuple[TupleType, OpResult]]
        self.reset_counter = 0

    def next(self, tup: TupleType):
        grouping_key = self.groupby(tup)
        hash_key = make_hashable(grouping_key)
        if hash_key in self.h_tbl:
            _, val = self.h_tbl[hash_key]
            new_val = self.reduce(val, tup)
            self.h_tbl[hash_key] = (grouping_key, new_val)
        else:
            new_val = self.reduce(Empty(), tup)
            self.h_tbl[hash_key] = (grouping_key, new_val)

    def reset(self, tup: TupleType):
        self.reset_counter += 1
        for grouping_key, val in self.h_tbl.values():
            unioned_tup = {**tup, **grouping_key}
            final_tup = {**unioned_tup, self.out_key: val}
            self.next_op.next(final_tup)
        self.next_op.reset(tup)
        self.h_tbl.clear()

def filter_groups(incl_keys: List[str]) -> Callable[[TupleType], TupleType]:
    def f(tup: TupleType) -> TupleType:
        return {k: v for k, v in tup.items() if k in incl_keys}
    return f

def single_group(_: TupleType) -> TupleType:
    return {}

def counter(val: OpResult, _: TupleType) -> OpResult:
    if isinstance(val, Empty):
        return Int(1)
    elif isinstance(val, Int):
        return Int(val.value + 1)
    return val

def sum_ints(search_key: str) -> Callable[[OpResult, TupleType], OpResult]:
    def f(init_val: OpResult, tup: TupleType) -> OpResult:
        if isinstance(init_val, Empty):
            return Int(0)
        elif isinstance(init_val, Int):
            if search_key in tup and isinstance(tup[search_key], Int):
                return Int(tup[search_key].value + init_val.value)
            raise ValueError(f"'sum_ints' failed to find integer value mapped to '{search_key}'")
        return init_val
    return f

class DistinctOperator(Operator):
    def __init__(self, groupby: Callable[[TupleType], TupleType], next_op: Operator):
        self.groupby = groupby
        self.next_op = next_op
        self.seen = {}  # Dict[tuple, TupleType]
        self.reset_counter = 0

    def next(self, tup: TupleType):
        grouping_key = self.groupby(tup)
        hash_key = make_hashable(grouping_key)
        if hash_key not in self.seen:
            self.seen[hash_key] = grouping_key

    def reset(self, tup: TupleType):
        self.reset_counter += 1
        for grouping_key in self.seen.values():
            merged_tup = {**tup, **grouping_key}
            self.next_op.next(merged_tup)
        self.next_op.reset(tup)
        self.seen.clear()

class SplitOperator(Operator):
    def __init__(self, left: Operator, right: Operator):
        self.left = left
        self.right = right

    def next(self, tup: TupleType):
        self.left.next(tup)
        self.right.next(tup)

    def reset(self, tup: TupleType):
        self.left.reset(tup)
        self.right.reset(tup)

def join(eid_key: str, left_extractor: Callable[[TupleType], Tuple[TupleType, TupleType]],
         right_extractor: Callable[[TupleType], Tuple[TupleType, TupleType]], next_op: Operator) -> Tuple[Operator, Operator]:
    eid_key = "eid" if eid_key == None else eid_key
    h_tbl1 = {}
    h_tbl2 = {}
    left_curr_epoch = 0
    right_curr_epoch = 0

    class JoinSideOperator(Operator):
        def __init__(self, curr_h_tbl, other_h_tbl, curr_epoch_ref, other_epoch_ref, f):
            self.curr_h_tbl = curr_h_tbl
            self.other_h_tbl = other_h_tbl
            self.curr_epoch_ref = [curr_epoch_ref]  # Mutable reference
            self.other_epoch_ref = [other_epoch_ref]
            self.f = f
            self.next_op = next_op

        def next(self, tup: TupleType):
            key, vals_ = self.f(tup)
            curr_epoch = get_mapped_int(eid_key, tup)
            while curr_epoch > self.curr_epoch_ref[0]:
                if self.other_epoch_ref[0] > self.curr_epoch_ref[0]:
                    self.next_op.reset({eid_key: Int(self.curr_epoch_ref[0])})
                self.curr_epoch_ref[0] += 1
            new_tup = {**key, eid_key: Int(curr_epoch)}
            hash_key = make_hashable(new_tup)
            if hash_key in self.other_h_tbl:
                val_ = self.other_h_tbl.pop(hash_key)
                self.next_op.next({**new_tup, **vals_, **val_})
            else:
                self.curr_h_tbl[hash_key] = vals_

        def reset(self, tup: TupleType):
            curr_epoch = get_mapped_int(eid_key, tup)
            while curr_epoch > self.curr_epoch_ref[0]:
                if self.other_epoch_ref[0] > self.curr_epoch_ref[0]:
                    self.next_op.reset({eid_key: Int(self.curr_epoch_ref[0])})
                self.curr_epoch_ref[0] += 1

    left_op = JoinSideOperator(h_tbl1, h_tbl2, left_curr_epoch, right_curr_epoch, left_extractor)
    right_op = JoinSideOperator(h_tbl2, h_tbl1, right_curr_epoch, left_curr_epoch, right_extractor)
    return left_op, right_op

def rename_filtered_keys(renamings_pairs: List[Tuple[str, str]]) -> Callable[[TupleType], TupleType]:
    def f(in_tup: TupleType) -> TupleType:
        new_tup = {}
        for old_key, new_key in renamings_pairs:
            if old_key in in_tup:
                new_tup[new_key] = in_tup[old_key]
        return new_tup
    return f

# Query definitions
def ident(next_op: Operator) -> Operator:
    def filter_func(tup: TupleType) -> TupleType:
        return {k: v for k, v in tup.items() if k != "eth.src" and k != "eth.dst"}
    return MapOperator(filter_func, next_op)

def count_pkts(next_op: Operator) -> Operator:
    return EpochOperator(1.0, "eid", GroupByOperator(single_group, counter, "pkts", next_op))

def pkts_per_src_dst(next_op: Operator) -> Operator:
    return EpochOperator(1.0, "eid", GroupByOperator(filter_groups(["ipv4.src", "ipv4.dst"]), counter, "pkts", next_op))

def distinct_srcs(next_op: Operator) -> Operator:
    return EpochOperator(1.0, "eid", DistinctOperator(filter_groups(["ipv4.src"]), GroupByOperator(single_group, counter, "srcs", next_op)))

def tcp_new_cons(next_op: Operator) -> Operator:
    threshold = 40
    def tcp_filter(tup: TupleType) -> bool:
        return get_mapped_int("ipv4.proto", tup) == 6 and get_mapped_int("l4.flags", tup) == 2
    return EpochOperator(1.0, "eid", FilterOperator(tcp_filter, GroupByOperator(filter_groups(["ipv4.dst"]), counter, "cons", FilterOperator(key_geq_int("cons", threshold), next_op))))

def ssh_brute_force(next_op: Operator) -> Operator:
    threshold = 40
    def ssh_filter(tup: TupleType) -> bool:
        return get_mapped_int("ipv4.proto", tup) == 6 and get_mapped_int("l4.dport", tup) == 22
    return EpochOperator(1.0, "eid", FilterOperator(ssh_filter, DistinctOperator(filter_groups(["ipv4.src", "ipv4.dst", "ipv4.len"]), GroupByOperator(filter_groups(["ipv4.dst", "ipv4.len"]), counter, "srcs", FilterOperator(key_geq_int("srcs", threshold), next_op)))))

def super_spreader(next_op: Operator) -> Operator:
    threshold = 40
    return EpochOperator(1.0, "eid", DistinctOperator(filter_groups(["ipv4.src", "ipv4.dst"]), GroupByOperator(filter_groups(["ipv4.src"]), counter, "dsts", FilterOperator(key_geq_int("dsts", threshold), next_op))))

def port_scan(next_op: Operator) -> Operator:
    threshold = 40
    return EpochOperator(1.0, "eid", DistinctOperator(filter_groups(["ipv4.src", "l4.dport"]), GroupByOperator(filter_groups(["ipv4.src"]), counter, "ports", FilterOperator(key_geq_int("ports", threshold), next_op))))

def ddos(next_op: Operator) -> Operator:
    threshold = 45
    return EpochOperator(1.0, "eid", DistinctOperator(filter_groups(["ipv4.src", "ipv4.dst"]), GroupByOperator(filter_groups(["ipv4.dst"]), counter, "srcs", FilterOperator(key_geq_int("srcs", threshold), next_op))))

def syn_flood_sonata(next_op: Operator) -> List[Operator]:
    threshold = 3
    epoch_dur = 1.0

    def syns_filter(tup: TupleType) -> bool:
        return get_mapped_int("ipv4.proto", tup) == 6 and get_mapped_int("l4.flags", tup) == 2
    def synacks_filter(tup: TupleType) -> bool:
        return get_mapped_int("ipv4.proto", tup) == 6 and get_mapped_int("l4.flags", tup) == 18
    def acks_filter(tup: TupleType) -> bool:
        return get_mapped_int("ipv4.proto", tup) == 6 and get_mapped_int("l4.flags", tup) == 16

    syns_op = EpochOperator(epoch_dur, "eid", FilterOperator(syns_filter, GroupByOperator(filter_groups(["ipv4.dst"]), counter, "syns", Operator())))
    synacks_op = EpochOperator(epoch_dur, "eid", FilterOperator(synacks_filter, GroupByOperator(filter_groups(["ipv4.src"]), counter, "synacks", Operator())))
    acks_op = EpochOperator(epoch_dur, "eid", FilterOperator(acks_filter, GroupByOperator(filter_groups(["ipv4.dst"]), counter, "acks", Operator())))

    join_op1, join_op2 = join("eid",
        left_extractor=lambda tup: (filter_groups(["host"])(tup), filter_groups(["syns+synacks"])(tup)),
        right_extractor=lambda tup: (rename_filtered_keys([("ipv4.dst", "host")])(tup), filter_groups(["acks"])(tup)),
        next_op=MapOperator(
            lambda tup: {**tup, "syns+synacks-acks": Int(get_mapped_int("syns+synacks", tup) - get_mapped_int("acks", tup))},
            FilterOperator(key_geq_int("syns+synacks-acks", threshold), next_op)
        )
    )
    join_op3, join_op4 = join("eid",
        left_extractor=lambda tup: (rename_filtered_keys([("ipv4.dst", "host")])(tup), filter_groups(["syns"])(tup)),
        right_extractor=lambda tup: (rename_filtered_keys([("ipv4.src", "host")])(tup), filter_groups(["synacks"])(tup)),
        next_op=MapOperator(
            lambda tup: {**tup, "syns+synacks": Int(get_mapped_int("syns", tup) + get_mapped_int("synacks", tup))},
            join_op1
        )
    )
    return [syns_op.next_op.next_op.next_op.next_op(join_op3), synacks_op.next_op.next_op.next_op.next_op(join_op4), acks_op.next_op.next_op.next_op.next_op(join_op2)]

def completed_flows(next_op: Operator) -> List[Operator]:
    threshold = 1
    epoch_dur = 30.0

    def syns_filter(tup: TupleType) -> bool:
        return get_mapped_int("ipv4.proto", tup) == 6 and get_mapped_int("l4.flags", tup) == 2
    def fins_filter(tup: TupleType) -> bool:
        return get_mapped_int("ipv4.proto", tup) == 6 and (get_mapped_int("l4.flags", tup) & 1) == 1

    syns_op = EpochOperator(epoch_dur, "eid", FilterOperator(syns_filter, GroupByOperator(filter_groups(["ipv4.dst"]), counter, "syns", Operator())))
    fins_op = EpochOperator(epoch_dur, "eid", FilterOperator(fins_filter, GroupByOperator(filter_groups(["ipv4.src"]), counter, "fins", Operator())))

    op1, op2 = join("eid",
        left_extractor=lambda tup: (rename_filtered_keys([("ipv4.dst", "host")])(tup), filter_groups(["syns"])(tup)),
        right_extractor=lambda tup: (rename_filtered_keys([("ipv4.src", "host")])(tup), filter_groups(["fins"])(tup)),
        next_op=MapOperator(
            lambda tup: {**tup, "diff": Int(get_mapped_int("syns", tup) - get_mapped_int("fins", tup))},
            FilterOperator(key_geq_int("diff", threshold), next_op)
        )
    )
    return [syns_op.next_op.next_op.next_op.next_op(op1), fins_op.next_op.next_op.next_op.next_op(op2)]

def slowloris(next_op: Operator) -> List[Operator]:
    t1, t2, t3 = 5, 500, 90
    epoch_dur = 1.0

    def tcp_filter(tup: TupleType) -> bool:
        return get_mapped_int("ipv4.proto", tup) == 6

    n_conns_op = EpochOperator(epoch_dur, "eid", FilterOperator(tcp_filter, DistinctOperator(filter_groups(["ipv4.src", "ipv4.dst", "l4.sport"]), GroupByOperator(filter_groups(["ipv4.dst"]), counter, "n_conns", FilterOperator(lambda tup: get_mapped_int("n_conns", tup) >= t1, Operator())))))
    n_bytes_op = EpochOperator(epoch_dur, "eid", FilterOperator(tcp_filter, GroupByOperator(filter_groups(["ipv4.dst"]), sum_ints("ipv4.len"), "n_bytes", FilterOperator(lambda tup: get_mapped_int("n_bytes", tup) >= t2, Operator()))))

    op1, op2 = join("eid",
        left_extractor=lambda tup: (filter_groups(["ipv4.dst"])(tup), filter_groups(["n_conns"])(tup)),
        right_extractor=lambda tup: (filter_groups(["ipv4.dst"])(tup), filter_groups(["n_bytes"])(tup)),
        next_op=MapOperator(
            lambda tup: {**tup, "bytes_per_conn": Int(get_mapped_int("n_bytes", tup) // get_mapped_int("n_conns", tup))},
            FilterOperator(lambda tup: get_mapped_int("bytes_per_conn", tup) <= t3, next_op)
        )
    )
    return [n_conns_op.next_op.next_op.next_op.next_op.next_op.next_op(op1), n_bytes_op.next_op.next_op.next_op.next_op.next_op(op2)]

# Main entry point
queries = [ident(DumpTupleOperator(sys.stdout))]

def run_queries():
    sample_tuples = [
        {
            "time": Float(0.0 + i),
            "eth.src": MAC(b"\x00\x11\x22\x33\x44\x55"),
            "eth.dst": MAC(b"\xAA\xBB\xCC\xDD\xEE\xFF"),
            "eth.ethertype": Int(0x0800),
            "ipv4.hlen": Int(20),
            "ipv4.proto": Int(6),
            "ipv4.len": Int(60),
            "ipv4.src": IPv4(ipaddress.IPv4Address("127.0.0.1")),
            "ipv4.dst": IPv4(ipaddress.IPv4Address("127.0.0.1")),
            "l4.sport": Int(440),
            "l4.dport": Int(50000),
            "l4.flags": Int(10)
        }
        for i in range(5)
    ]
    for tup in sample_tuples:
        for query in queries:
            query.next(tup)

if __name__ == "__main__":
    run_queries()
    print("Done")

File: LLM-linguistic-translations/OpenAi-chatgpt/4o/cpp-translations/chat-code.cpp
#include <iostream>
#include <iomanip>
#include <string>
#include <map>
#include <unordered_map>
#include <variant>
#include <vector>
#include <array>
#include <sstream>
#include <functional>
#include <set>
#include <memory>
#include <stdexcept>

class IPv4Addr {
public:
    explicit IPv4Addr(const std::string& addr) : address(addr) {}
    std::string to_string() const { return address; }
    bool operator==(const IPv4Addr& o) const { return address == o.address; }
    bool operator<(const IPv4Addr& o) const { return address < o.address; }
private:
    std::string address;
};

using MACAddr = std::array<uint8_t, 6>;

using OpResult = std::variant<std::monostate, int, float, IPv4Addr, MACAddr>;
using Tuple = std::map<std::string, OpResult>;

std::string to_string(const OpResult& val) {
    struct {
        std::string operator()(std::monostate) const { return "Empty"; }
        std::string operator()(int i) const { return std::to_string(i); }
        std::string operator()(float f) const {
            std::ostringstream oss; oss << std::fixed << std::setprecision(6) << f; return oss.str();
        }
        std::string operator()(const IPv4Addr& ip) const { return ip.to_string(); }
        std::string operator()(const MACAddr& mac) const {
            std::ostringstream oss;
            for (size_t i = 0; i < 6; ++i) {
                if (i > 0) oss << ":";
                oss << std::hex << std::setw(2) << std::setfill('0') << static_cast<int>(mac[i]);
            }
            return oss.str();
        }
    } visitor;
    return std::visit(visitor, val);
}

int get_int(const OpResult& v) {
    if (auto p = std::get_if<int>(&v)) return *p;
    throw std::runtime_error("Expected int");
}

float get_float(const OpResult& v) {
    if (auto p = std::get_if<float>(&v)) return *p;
    throw std::runtime_error("Expected float");
}

std::string tuple_to_string(const Tuple& tup) {
    std::ostringstream oss;
    for (const auto& [k, v] : tup)
        oss << "\"" << k << "\" => " << to_string(v) << ", ";
    return oss.str();
}

void print_tuple(const Tuple& tup) {
    std::cout << tuple_to_string(tup) << "\n";
}

Tuple merge_tuples(const Tuple& a, const Tuple& b) {
    Tuple result = a;
    result.insert(b.begin(), b.end());
    return result;
}


struct Operator {
    virtual void next(const Tuple&) = 0;
    virtual void reset(const Tuple&) = 0;
    virtual ~Operator() = default;
};

using OperatorPtr = std::shared_ptr<Operator>;
using OpCreator = std::function<OperatorPtr(OperatorPtr)>;
using DblOpCreator = std::function<std::pair<OperatorPtr, OperatorPtr>(OperatorPtr)>;

OperatorPtr operator>>(OpCreator a, OperatorPtr b) { return a(b); }
std::pair<OperatorPtr, OperatorPtr> operator>>(DblOpCreator a, OperatorPtr b) { return a(b); }


class DumpTuple : public Operator {
public:
    explicit DumpTuple(bool show = false) : showReset(show) {}
    void next(const Tuple& t) override { print_tuple(t); }
    void reset(const Tuple& t) override {
        if (showReset) {
            print_tuple(t);
            std::cout << "[reset]\n";
        }
    }
private:
    bool showReset;
};

class FilterMap : public Operator {
public:
    explicit FilterMap(OperatorPtr next) : nextOp(std::move(next)) {}
    void next(const Tuple& t) override {
        Tuple filtered;
        for (const auto& [k, v] : t)
            if (k != "eth.src" && k != "eth.dst")
                filtered[k] = v;
        nextOp->next(filtered);
    }
    void reset(const Tuple& t) override { nextOp->reset(t); }
private:
    OperatorPtr nextOp;
};

OpCreator ident = [](OperatorPtr next) {
    return std::make_shared<FilterMap>(next);
};

class Epoch : public Operator {
public:
    Epoch(float width, std::string key, OperatorPtr next)
        : epochWidth(width), keyOut(std::move(key)), nextOp(std::move(next)) {}

    void next(const Tuple& tup) override {
        float time = get_float(tup.at("time"));
        if (epochBoundary == 0.0f) epochBoundary = time + epochWidth;
        else if (time >= epochBoundary) {
            while (time >= epochBoundary) {
                Tuple resetTup = {{keyOut, epochId}};
                nextOp->reset(resetTup);
                epochBoundary += epochWidth;
                ++epochId;
            }
        }
        Tuple out = tup;
        out[keyOut] = epochId;
        nextOp->next(out);
    }

    void reset(const Tuple&) override {
        nextOp->reset({{keyOut, epochId}});
        epochBoundary = 0.0f;
        epochId = 0;
    }

private:
    float epochWidth, epochBoundary = 0.0f;
    int epochId = 0;
    std::string keyOut;
    OperatorPtr nextOp;
};

OpCreator epoch(float width, std::string key) {
    return [=](OperatorPtr next) {
        return std::make_shared<Epoch>(width, key, next);
    };
}

class Filter : public Operator {
public:
    using Predicate = std::function<bool(const Tuple&)>;
    Filter(Predicate pred, OperatorPtr next) : pred(std::move(pred)), nextOp(std::move(next)) {}
    void next(const Tuple& tup) override { if (pred(tup)) nextOp->next(tup); }
    void reset(const Tuple& tup) override { nextOp->reset(tup); }
private:
    Predicate pred;
    OperatorPtr nextOp;
};

OpCreator filter(Filter::Predicate pred) {
    return [=](OperatorPtr next) {
        return std::make_shared<Filter>(pred, next);
    };
}

class MapOp : public Operator {
public:
    using Mapper = std::function<Tuple(const Tuple&)>;
    MapOp(Mapper m, OperatorPtr next) : mapFn(std::move(m)), nextOp(std::move(next)) {}
    void next(const Tuple& tup) override { nextOp->next(mapFn(tup)); }
    void reset(const Tuple& tup) override { nextOp->reset(tup); }
private:
    Mapper mapFn;
    OperatorPtr nextOp;
};

OpCreator map(MapOp::Mapper fn) {
    return [=](OperatorPtr next) {
        return std::make_shared<MapOp>(fn, next);
    };
}

int get_mapped_int(const std::string& key, const Tuple& tup) {
    return get_int(tup.at(key));
}

bool key_geq_int(const std::string& key, int threshold, const Tuple& tup) {
    return get_mapped_int(key, tup) >= threshold;
}

std::function<Tuple(const Tuple&)> filter_keys(const std::vector<std::string>& keys) {
    return [=](const Tuple& tup) {
        Tuple out;
        for (const auto& key : keys) {
            auto it = tup.find(key);
            if (it != tup.end()) out[key] = it->second;
        }
        return out;
    };
}

int main() {
    OperatorPtr printer = std::make_shared<DumpTuple>(true);
    OperatorPtr pipeline = epoch(1.0, "eid") >> ident >> printer;

    for (int i = 0; i < 20; ++i) {
        Tuple tup;
        tup["time"] = static_cast<float>(i);
        tup["eth.src"] = MACAddr{0x00, 0x11, 0x22, 0x33, 0x44, 0x55};
        tup["eth.dst"] = MACAddr{0xAA, 0xBB, 0xCC, 0xDD, 0xEE, 0xFF};
        tup["eth.ethertype"] = 0x0800;
        tup["ipv4.hlen"] = 20;
        tup["ipv4.proto"] = 6;
        tup["ipv4.len"] = 60;
        tup["ipv4.src"] = IPv4Addr("127.0.0.1");
        tup["ipv4.dst"] = IPv4Addr("127.0.0.1");
        tup["l4.sport"] = 440;
        tup["l4.dport"] = 50000;
        tup["l4.flags"] = 10;

        pipeline->next(tup);
    }

    pipeline->reset({});
    std::cout << "Done\n";
    return 0;
}
